main/JSONParameterUtilsMod.F90

```diff
@@ -0,0 +1,1304 @@
+                                                        ! combines to explain dimensionality
+                                                        ! and type of value, such as int/string/float
+     character(len=max_sl), allocatable :: dim_names(:) ! These are the indices of the dimensions
+                                                        ! associated with this variable
+     integer               :: ndims                     ! Number of dimensions,
+                                                        ! same as size (dim_names)
+     integer                         :: access_count    ! This is used to count how many
+                                                        ! times this parameter has been read
+                                                        ! on initialization, expecte number
+                                                        ! of times is 1, helps for debugging
+
+     ! --- Data holding structures. Each parameter will only use ONE of these ----
+     real(r8)                        :: r_data_scalar
+     real(r8), allocatable           :: r_data_1d(:)
+     real(r8), allocatable           :: r_data_2d(:,:)
+     integer                         :: i_data_scalar
+     integer,  allocatable           :: i_data_1d(:)
+     integer,  allocatable           :: i_data_2d(:,:)
+     character(len=128)              :: c_data
+     character(len=128), allocatable :: c_data_1d(:)
+     character(len=128), allocatable :: c_data_2d(:,:)
+     
+  end type param_type
+
+  type,public :: params_type
+     type(param_type), pointer :: parameters(:)
+     type(dim_type), pointer :: dimensions(:)
+   contains
+     procedure :: GetDimSizeFromName
+     procedure :: GetParamFromName
+     procedure :: ReportAccessCounts
+  end type params_type
+  
+  public :: ReadJSON
+  public :: SetInvalid
+  public :: SetLogInit
+  public :: DumpParameter
+  
+contains
+
+  subroutine SetInvalid(r_invalid_in)
+    real(r8),intent(in) :: r_invalid_in
+    r_invalid = r_invalid_in
+  end subroutine SetInvalid
+  
+  subroutine SetLogInit(log_unit_in)
+    integer,intent(in) :: log_unit_in
+    log_unit = log_unit_in
+  end subroutine SetLogInit
+
+  ! ==============================================================================
+  
+  subroutine ReadJSON(filename,file_unit,pstruct)
+
+    ! This is esssentially the driver for reading through the input parameter file
+    ! There should be two meaningful sections (not including attributes)
+    ! 1) dimensions
+    ! 2) parameters
+    !
+    ! All parameters must be scalar or associated with a known dimension.
+    !
+    ! Each section is started with an open bracket '{' and closed with a closed bracket '}'
+    ! For each section here we end with a call
+    ! to rewind and go back to the top of the file
+    
+    character(len=*),intent(in) :: filename
+    integer,intent(in)          :: file_unit
+    type(params_type)           :: pstruct
```

Comment:
```suggestion
    type(params_type), intent(inout)           :: pstruct
```

Reply:
It's good to always have an intent in arguments to show you thought about it. It does look like this needs to be both IN and OUT, if I'm reading it right. Since, it's a read method, making it a OUT could be a good thing to try for.